common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,1056 @@
+  MetadataGroup g;
+#if CARBON_NEON_SIMD_SUPPORT
+  g.metadata_vec = vld1_u8(metadata + index);
+#elif CARBON_X86_SIMD_SUPPORT
+  g.metadata_vec =
+      _mm_load_si128(reinterpret_cast<const __m128i*>(metadata + index));
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+  return g;
+}
+
+inline auto MetadataGroup::SIMDStore(uint8_t* metadata, ssize_t index) const
+    -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  vst1_u8(metadata + index, metadata_vec);
+#elif CARBON_X86_SIMD_SUPPORT
+  _mm_store_si128(reinterpret_cast<__m128i*>(metadata + index), metadata_vec);
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+}
+
+inline auto MetadataGroup::SIMDClearDeleted() -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  // There is no good Neon operation to implement this, so do it using integer
+  // code. This is reasonably fast, but unfortunate because it forces the group
+  // out of a SIMD register and into a general purpose register, which can have
+  // high latency.
+  metadata_ints[0] &= (~LSBs | metadata_ints[0] >> 7);
+#elif CARBON_X86_SIMD_SUPPORT
+  metadata_vec =
+      _mm_blendv_epi8(_mm_setzero_si128(), metadata_vec, metadata_vec);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+}
+
+inline auto MetadataGroup::SIMDMatch(uint8_t tag) const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto match_byte_vec = vdup_n_u8(tag | PresentMask);
+  auto match_byte_cmp_vec = vceq_u8(metadata_vec, match_byte_vec);
+  uint64_t match_bits = vreinterpret_u64_u8(match_byte_cmp_vec)[0];
+  // The matched range is likely to be tested for zero by the caller, and that
+  // test can often be folded into masking the bits with `MSBs` when we do that
+  // mask in the scalar domain rather than the SIMD domain. So we do the mask
+  // here rather than above prior to extracting the match bits.
+  result = MatchRange(match_bits & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = X86SIMDMatch(tag | PresentMask);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(tag);
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchPresent() const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  // Just extract directly.
+  uint64_t match_bits = vreinterpret_u64_u8(metadata_vec)[0];
+  // The matched range is likely to be tested for zero by the caller, and that
+  // test can often be folded into masking the bits with `MSBs` when we do that
+  // mask in the scalar domain rather than the SIMD domain. So we do the mask
+  // here rather than above prior to extracting the match bits.
+  result = MatchRange(match_bits & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  // We arrange the byte vector for present bytes so that we can directly
+  // extract it as a mask.
```

Comment:
Less sure of this one
```suggestion
  // extract it as match bits.
```